cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,361 @@
+                    total lived time of the lineage, and the status of whether
+                    the lineage is still dividing
+
+            Returns:
+                The new unique ID to use for the next event, incremented only if
+                a new descendant is added to the tree
+            """
+            assert(lineage["active"])
+
+            unique_id = next(names)
+
+            birth_waiting_time = self.birth_waiting_distribution(lineage["birth_scale"])
+            death_waiting_time = self.death_waiting_distribution()
+            if birth_waiting_time <= 0 or death_waiting_time <= 0:
+                raise TreeSimulatorError(
+                    "0 or negative waiting time detected"
+                )
+
+            # If birth and death would happen after the total experiment time,
+            # just cut off the living branch length at the experiment time
+            if (
+                self.experiment_time
+                and lineage["total_time"] + birth_waiting_time
+                >= self.experiment_time
+                and lineage["total_time"] + death_waiting_time
+                >= self.experiment_time
+            ):
+                tree.add_node(unique_id)
+                tree.nodes[unique_id]["birth_scale"] = lineage["birth_scale"]
+                tree.add_edge(
+                    lineage["id"],
+                    unique_id,
+                    weight=self.experiment_time - lineage["total_time"],
+                )
+                tree.nodes[unique_id]["total_time"] = self.experiment_time
+
+                current_lineages.put(
+                    (
+                        self.experiment_time,
+                        unique_id,
+                        {
+                            "id": unique_id,
+                            "birth_scale": lineage["birth_scale"],
+                            "total_time": self.experiment_time,
+                            "active": False,
+                        },
+                    )
+                )
+                observed_nodes.append(unique_id)
+
+            else:
+                if birth_waiting_time < death_waiting_time:
+                    # Update fitness
+                    base_selection_coefficient = 1
+                    if self.mutation_distribution:
+                        num_mutations = int(self.mutation_distribution())
+                        if num_mutations < 0:
+                            raise TreeSimulatorError(
+                                "Negative number of mutations detected"
+                            )
+                        for _ in range(num_mutations):
+                            base_selection_coefficient *= self.fitness_distribution()
+                            if base_selection_coefficient < 0:
+                                raise TreeSimulatorError(
+                                    "Negative mutation strength detected"
+                                )
```

Comment:
Let's move the logic of computing the new birth scale parameter to a separate function; it will make the code more readable and easier to refactor if we want to change the way that birth scales are updated.